db/db_compaction_test.cc

```diff
@@ -5109,6 +5109,52 @@ TEST_F(DBCompactionTest, CompactionDuringShutdown) {
   ASSERT_OK(dbfull()->error_handler_.GetBGError());
 }
 
+TEST_F(DBCompactionTest, CancelSkippingBottomTombstones) {
+  Options opts = CurrentOptions();
+  opts.disable_auto_compactions = true;
+  opts.num_levels = 3;
+  opts.level0_file_num_compaction_trigger = 32;
+  const int kNumL0Files = opts.level0_file_num_compaction_trigger;
+  DestroyAndReopen(opts);
+
+  // Insert a random value
+  ASSERT_OK(Put("b", "value"));
+
+  // Take a snapshot
+  const Snapshot* snapshot = db_->GetSnapshot();
+
+  for (int i = 0; i < kNumL0Files / 4; ++i) {
+    ASSERT_OK(Put("a", "value"));
+    ASSERT_OK(Put("c", "value"));
+    ASSERT_OK(Flush());
+
+    ASSERT_OK(Delete("a"));
+    ASSERT_OK(Flush());
+
+    ASSERT_OK(Delete("a"));
+    ASSERT_OK(Flush());
+  }
+
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->LoadDependency({
+      {"CompactionIterator:ProcessKV",
+       "DBImpl::CancelAllBackgroundWork:BeforeSetShutdownFlag"},
+      {"DBImpl::CancelAllBackgroundWork:AfterSetShutdownFlag",
+       "CompactionIterator::NextFromInput:BeforeSkippingBottomTombstones:1"},
+  });
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  port::Thread thread([&]() { dbfull()->CancelAllBackgroundWork(true); });
+
+  CompactRangeOptions cro;
+  Status s = db_->CompactRange(cro, nullptr, nullptr);
+  ASSERT_TRUE(s.IsShutdownInProgress());
```

Comment:
I'm wondering if we can somehow make sure that in this test that the new logic kicked in? (I believe you would get the same behavior with the old code as well once we've churned through the tombstones.) Maybe instead of changing the condition of the `while` above, we could put a check with a sync point in the `while` loop's body, e.g. `if (IsShuttingDown()) { TEST_SYNC_POINT(...); break; }`?